src/ucp/core/ucp_ep.c

```diff
@@ -3475,11 +3475,9 @@ static void ucp_ep_req_purge_send(ucp_request_t *req, ucs_status_t status)
     ucs_assertv(UCS_STATUS_IS_ERR(status), "req %p: status %s", req,
                 ucs_status_string(status));
 
-    if ((ucp_ep_config(req->send.ep)->key.err_mode !=
-         UCP_ERR_HANDLING_MODE_NONE) &&
-        (req->flags & UCP_REQUEST_FLAG_RKEY_INUSE) &&
-        !(req->flags & UCP_REQUEST_FLAG_USER_MEMH)) {
+    if (ucp_request_invalidation_required(req)) {
         ucp_request_dt_invalidate(req, status);
+        ucp_memh_put(req->send.state.dt.dt.contig.memh);
```

Comment:
after recent changes, it makes more sense to move ucp_memh_put inside ucp_request_memh_invalidate.
because it's not good to use the success on invalidate as indirect indication that data type is contain

Reply:
Done.